### PR TITLE
gh-92886: fixed tests that were breaking while running with basic optimizations enabled (-O, assertions off)

### LINKED ISSUE #92886: some tests fail with -O (assertions off)
On main (3.12 branch):

% ./python.exe -Om test -j2
[...]
10 tests failed:
    test_clinic test_coroutines test_dis test_imaplib
    test_multiprocessing_forkserver test_multiprocessing_spawn
    test_py_compile test_sys_settrace test_wsgiref test_zipimport

This seems wrong because
(1) it indicates that some assertions should be exceptions instead. 
(2) it implies that we don't have a buildbot testing with -O

- [x] test_clinic
- [x] test_coroutines
- [x] test_dis
- [x] test_imaplib
- [x] test_multiprocessing_forkserver
- [x] test_multiprocessing_spawn
- [x] test_py_compile
- [x] test_sys_settrace
- [x] test_wsgiref
- [x] test_zipimport

### PR DESCRIPTION
**Pull request by jackh-ncl**
Resolves #92886 

```
$ uname -a
Darwin ... 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:47:26 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T8101 arm64

$ ./python.exe -Om test -j8
...
== Tests result: SUCCESS ==

407 tests OK.

29 tests skipped:
    test_curses test_dbm_gnu test_devpoll test_epoll test_gdb
    test_idle test_ioctl test_launcher test_msilib
    test_multiprocessing_fork test_ossaudiodev test_smtpnet
    test_socketserver test_spwd test_ssl test_startfile test_tcl
    test_tix test_tk test_ttk_guionly test_ttk_textonly test_turtle
    test_urllib2net test_urllibnet test_winconsoleio test_winreg
    test_winsound test_xmlrpc_net test_zipfile64

Total duration: 3 min 49 sec
Tests result: SUCCESS
```

### CODE REVIEW

File: Lib/test/_test_multiprocessing.py [hunk 1]
```diff
@@ -5690,45 +5690,48 @@ def test_joinable_queue(self):
```
Comment (jackh-ncl) [inline Lib/test/_test_multiprocessing.py:5690]: Could be worth fixing the rest of the assertions in this file?

File: Lib/test/test_dis.py [hunk 2]
```diff
@@ -1041,11 +1041,6 @@ def get_disassembly(self, func, lasti=-1, wrapper=True, **kwargs):
         return output.getvalue()
 
 
-if sys.flags.optimize:
-    code_info_consts = "0: None"
-else:
-    code_info_consts = "0: 'Formatted details of methods, functions, or code.'"
-
```
Comment (jackh-ncl) [inline Lib/test/test_dis.py:1048]: Failing tests were:

image

It seemed as though this condition was not necessary anymore, though I'm unsure if it is the expected behaviour.

File: Lib/test/test_imaplib.py [hunk 3]
```diff
@@ -939,6 +939,7 @@ def test_with_statement_logout(self):
 
     @threading_helper.reap_threads
     @cpython_only
+    @unittest.skipUnless(__debug__, "Won't work if __debug__ is False")
```
Comment (jackh-ncl) [inline Lib/test/test_imaplib.py:942]: `_mesg()` @ `Lib/imaplib.py:1247` is defined within a `if __debug__` block, so I assume this test can be skipped when running with optimizations.

File: Lib/test/test_py_compile.py [hunk 4]
```diff
@@ -235,11 +235,12 @@ def pycompilecmd(self, *args, **kwargs):
         # assert_python_* helpers don't return proc object. We'll just use
         # subprocess.run() instead of spawn_python() and its friends to test
         # stdin support of the CLI.
+        opts = '-m' if __debug__ else '-Om'
```
Comment (jackh-ncl) [inline Lib/test/test_py_compile.py:238]: I'm not sure how this actually fixed things, but it seems to have done the trick 🤷‍♂️

File: Lib/test/test_sys_settrace.py [hunk 5]
```diff
@@ -833,9 +833,8 @@ def func():
              (5, 'line'),
              (6, 'line'),
              (7, 'line'),
-             (10, 'line'),
-             (13, 'line'),
-             (13, 'return')])
+             (10, 'line')] +
+             ([(13, 'line'), (13, 'return')] if __debug__ else [(10, 'return')]))
```
Comment (jackh-ncl) [inline Lib/test/test_sys_settrace.py:837]: The assertions on line 819 affect the line numbers we are testing here when running with optimizations.
An alternative could be for these test functions to do something other than an `assert`.

**Comment by jackh-ncl** [comment]
```
$ ./python.exe -Om test -j8
...
== Tests result: SUCCESS ==

407 tests OK.

29 tests skipped:
    test_curses test_dbm_gnu test_devpoll test_epoll test_gdb
    test_idle test_ioctl test_launcher test_msilib
    test_multiprocessing_fork test_ossaudiodev test_smtpnet
    test_socketserver test_spwd test_ssl test_startfile test_tcl
    test_tix test_tk test_ttk_guionly test_ttk_textonly test_turtle
    test_urllib2net test_urllibnet test_winconsoleio test_winreg
    test_winsound test_xmlrpc_net test_zipfile64

Total duration: 3 min 49 sec
Tests result: SUCCESS
```

**Comment by erlend-aasland** [comment]
Please limit this PR to tests only (files in the Lib/test directory). It _may_ make sense to break this up in multiple PRs.

**Comment by jackh-ncl** [comment]
@erlend-aasland I've split it out into several PRs, which hopefully doesn't seem too overkill. I can see there being some level of back and forth for a number of them.

**Comment by erlend-aasland** [comment]
> @erlend-aasland I've split it out into several PRs, which hopefully doesn't seem too overkill. I can see there being some level of back and forth for a number of them.

Thanks. One PR per file was perhaps a little bit too much, but IMO it is better than one PR for everything.

IIUC, this PR is now superfluous and can be closed?

**Pull request closed without merging**